go/vt/vtgate/buffer/buffer.go

```diff
@@ -233,6 +162,15 @@ func (b *Buffer) ProcessPrimaryHealth(th *discovery.TabletHealth) {
 	sb.recordExternallyReparentedTimestamp(timestamp, th.Tablet.Alias)
 }
 
+func (b *Buffer) HandleKeyspaceEvent(ksevent *discovery.KeyspaceEvent) {
+	for _, shard := range ksevent.Shards {
+		sb := b.getOrCreateBuffer(shard.Target.Keyspace, shard.Target.Shard)
+		if sb != nil {
+			sb.recordKeyspaceEvent(shard.Tablet, shard.Serving)
```

Comment:
We actually don't trigger any buffering operations when we receive the healcheck fail for a primary! That's because these healthchecks are processed by the topology engine, which often lags behind the actual availability issue. Instead, what we do it start buffering once the `vtgate` itself fails to reach the primary (i.e. when we get an error return during a request), and the Keyspace Events handler is designed to detect when the availability incident is over _cluster-wide_.

Reply:
Thanks @vmg that makes sense. 
> start buffering once the vtgate itself fails to reach the primary (i.e. when we get an error return during a request)

it looks like the retry is triggered on these 3 kinds of error code: https://github.com/vitessio/vitess/blob/main/go/vt/vttablet/queryservice/wrapped.go#L75

My current understanding is vttablet maps to those error code here: [Code_FAILED_PRECONDITION](https://github.com/vitessio/vitess/blob/61ecf0d30c5be5e98f1465b86d0397339b66e52e/go/vt/vttablet/tabletserver/tabletserver.go#L1432) here, Code_UNAVAILABLE [here](https://github.com/vitessio/vitess/blob/61ecf0d30c5be5e98f1465b86d0397339b66e52e/go/vt/vttablet/tabletserver/tabletserver.go#L1421) and Code_CLUSTER_EVENT [here](https://github.com/vitessio/vitess/blob/61ecf0d30c5be5e98f1465b86d0397339b66e52e/go/vt/vttablet/tabletserver/tabletserver.go#L1434) - they seem not related to a primary failure. The other place where we return `Code_CLUSTER_EVENT` is when the primary tablet has !serving servingState, which only happen _after_ a reparent event.

I think I'm missing some details here on the failover behavior, could you shed some lights on it? Thanks

Reply:
The error handling on the tablet was wired up by @harshit-gangal, so I don't know all the details, but the actual buffering code is triggered with this check, so the only relevant error code when it comes to buffering is `CLUSTER_EVENT`:

https://github.com/vitessio/vitess/blob/ca5a27ed5aacbb523d678d910cda8d5571e653f0/go/vt/vtgate/buffer/buffer.go#L72-L75

Reply:
...In retrospect, this function could use a different name since `CLUSTER_EVENT` can also be caused by a resharding operation. :sweat_smile:

Reply:
Got it, looking at the check [here](https://github.com/vitessio/vitess/blob/main/go/vt/vttablet/tabletserver/state_manager.go#L351), looks like the buffering logic only covers _after_ a reparent (but before vtgate get notified via health check), a vttablet would return Code_CLUSTER_EVENT. Since `replHealthy` is always true for primary tablet and serving state only changes via a reparent event. Does that sound right to you?